### PR TITLE
Azure Monitor: adjust grouping logic and avoid duplicating documents to make the metricset TSDB-friendly

### PR DESCRIPTION
**Pull request by zmoog**
This is a rough draft for a revision of the Azure Metrics grouping logic to make it more TSDB friendly.

## Overview

(WHAT) Here is a summary of the changes introduced with this PR.

- Update the metric grouping logic
- Track metrics collection info
- Adjust collection interval 

### Update the metric grouping logic

Streamlines the metrics grouping logic to include all the fields the TSDB team identified as dimensions for the Azure Metrics events. 

Here are the current components of the grouping key:

- timestamp
- namespace
- resource ID
- resource Sub ID
- dimensions
- time grain

It also tries to make the grouping simpler to read. 

WHY: 

When TSDB is enabled, it drops events with the same timestamp and dimensions. The metricset must group all metrics values by timestamp+dimensions and create one event for each group to avoid data loss.

### Track metrics collection info

The metricset tracks the timestamp and time grain for each metrics collection. At the beginning of each iteration, it skips collecting a value if the metricset has already collected a value for the (time grain) period.

(WHY)

The metricset usually collects one data point for each collection period. When the time grain is larger than the collection period, the metricset collects the identical data point multiple times.

For example, consider a `PT1H` (one hour) time grain and a collection period of five minutes: without tracking, the metrics would collect the identical `PT1H` data point 12 times.

### Adjust collection interval 

Change the collection interval to `[{-2 x INTERVAL},{-1 x INTERVAL})` with a delay of `{INTERVAL}`. 

(WHY)

The collection interval was [2x the collection period or time grain](https://github.com/elastic/beats/blob/ed34c37f59c7bc0cf9e8051f7b5327c861b59467/x-pack/metricbeat/module/azure/client.go#L110-L116). This interval is too large, and we collected multiple data points for the same metric. There was some code to drop the additional data points, but it wasn't working in all cases. 

Glossary:
- collection interval: the time range used to fetch metrics values.
- collection period: time between metric collections (e.g., with a 5 min period, the metricset collects new metrics every 5 minutes)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] ~~I have made corresponding change to the default configuration files~~
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

- Relates to:
  - https://github.com/elastic/integrations/issues/7646
  - https://github.com/elastic/integrations/issues/7160
  - https://github.com/elastic/integrations/issues/7621

### CODE REVIEW

File: x-pack/metricbeat/module/azure/client.go [hunk 1]
```diff
@@ -97,24 +171,51 @@ func (client *Client) InitResources(fn mapResourceMetrics) error {
...
-		}
-		endTime := time.Now().UTC()
-		startTime := endTime.Add(interval * (-2))
+
+		// copy the reference time
+		endTime := referenceTime
+
+		// Fetch in the range [{-2 x INTERVAL},{-1 x INTERVAL}) with a delay of {INTERVAL}.
+		endTime = endTime.Add(interval * (-1))
+		startTime := endTime.Add(interval * (-1))
```
Comment (kaiyan-sheng) [inline x-pack/metricbeat/module/azure/client.go:193]: Ln 186 to 194 can all move out of the for loop I think?

Comment (zmoog) [reply x-pack/metricbeat/module/azure/client.go:193]: Yes, you're right!

I was convinced that method `Add()` changed the variable's value, but it's false.

I am moving this code block outside the for loop.

**Review by kaiyan-sheng — APPROVED** [review]

**Comment by tetianakravchenko** [comment]
Hey @zmoog should it be backported to the older versions? As I remember we discussed that those changes should be considered as a bug, not as a new feature, isn't it?

**Comment by zmoog** [comment]
Yeah, I remember the conversation about backpointing it to 8.11. Two of the three changes are bug fixes; the remaining is a tweak of an existing feature (metrics grouping).

So, it qualifies for a backport to 8.11. 

Shall we proceed?

**Pull request merged**